src/main/java/io/cdap/plugin/gcp/bigtable/source/BigtableSourceConfig.java

```diff
@@ -0,0 +1,206 @@
+  public static final String INSTANCE = "instance";
+  public static final String KEY_ALIAS = "keyAlias";
+  public static final String SCAN_ROW_START = "scanRowStart";
+  public static final String SCAN_ROW_STOP = "scanRowStop";
+  public static final String SCAN_TIME_RANGE_START = "scanTimeRangeStart";
+  public static final String SCAN_TIME_RANGE_STOP = "scanTimeRangeStop";
+  public static final String SCHEMA = "schema";
+  public static final String ON_ERROR = "on-error";
+
+  @Name(TABLE)
+  @Macro
+  @Description("The table to read from. A table contains individual records organized in rows. "
+    + "Each record is composed of columns (also called fields). "
+    + "Every table is defined by a schema that describes the column names, data types, and other information.")
+  final String table;
+
+  @Name(INSTANCE)
+  @Macro
+  @Description("BigTable instance id. " +
+    "Uniquely identifies BigTable instance within your Google Cloud Platform project.")
+  final String instance;
+
+  @Name(KEY_ALIAS)
+  @Description("Name of the field for row key.")
+  @Macro
+  @Nullable
+  final String keyAlias;
+
+  @Name(SCAN_ROW_START)
+  @Description("Scan start row.")
+  @Macro
+  @Nullable
+  final String scanRowStart;
+
+  @Name(SCAN_ROW_STOP)
+  @Description("Scan stop row.")
+  @Macro
+  @Nullable
+  final String scanRowStop;
+
+  @Name(SCAN_TIME_RANGE_START)
+  @Description("The starting timestamp used to filter columns with a specific range of versions.")
+  @Macro
+  @Nullable
+  final Long scanTimeRangeStart;
+
+  @Name(SCAN_TIME_RANGE_STOP)
+  @Description("The ending timestamp used to filter columns with a specific range of versions.")
+  @Macro
+  @Nullable
+  final Long scanTimeRangeStop;
+
+  @Name(ON_ERROR)
+  @Description("How to handle error in record processing. " +
+    "Error will be thrown if failed to parse value according to provided schema.")
+  @Macro
+  final String onError;
+
+  @Name(SCHEMA)
+  @Macro
+  @Description("The schema of the table to read.")
+  final String schema;
+
+  public BigtableSourceConfig(String table, String instance, @Nullable String project,
+                              @Nullable String serviceFilePath, @Nullable String keyAlias,
+                              @Nullable String scanRowStart, @Nullable String scanRowStop,
+                              @Nullable Long scanTimeRangeStart, @Nullable Long scanTimeRangeStop,
+                              String onError, String schema) {
+    this.table = table;
+    this.instance = instance;
+    this.project = project;
+    this.serviceFilePath = serviceFilePath;
+    this.keyAlias = keyAlias;
+    this.scanRowStart = scanRowStart;
+    this.scanRowStop = scanRowStop;
+    this.scanTimeRangeStart = scanTimeRangeStart;
+    this.scanTimeRangeStop = scanTimeRangeStop;
+    this.onError = onError;
+    this.schema = schema;
+  }
+
+  public void validate() {
+    super.validate();
+    if (!containsMacro(PROJECT) && null == project) {
```

Comment:
this is not correct, it should call getProject() to actually try and auto-detect.

Reply:
fixed but had to introduce new method to GCPConfig.